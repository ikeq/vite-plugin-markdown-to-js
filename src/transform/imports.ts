import camelCase from 'lodash/camelCase';
import { LangTransform } from '../parse';

const rSplit = /[\n\s,]/;
const rExt = /\.\w+$/;

/**
 * @example
 * ```md
 * ```imports
 * module1, module2, module3
 * module4
 * ```
 * ```
 */
export function transformImports({
  base,
  defaultPrefix,
}: { base?: string | ((path: string) => string); defaultPrefix?: string } = {}): LangTransform {
  const baseFn = (() => {
    if (typeof base === 'function') return base;
    if (typeof base === 'string') {
      if (base && !base.endsWith('/')) {
        base = `${base}/`;
      }
      return (path: string) => `${base || ''}${path}`;
    }
    return (path: string) => path;
  })();

  return {
    lang: 'imports',
    transform(src) {
      const imports = [...new Set(src.split(rSplit).filter((i) => i))];
      const importNames = imports.map((i) =>
        defaultPrefix ? camelCase(`${defaultPrefix} ${i}`) : i
      );

      return [
        {
          type: 'imports',
          imports: importNames,
          code: '',
        },
        {
          type: 'script',
          imports: importNames,
          code: imports
            .map((i, index) => {
              const path = baseFn(`${i}${rExt.test(i) ? '' : '.md'}`);
              return `import ${importNames[index]} from "${path}";`;
            })
            .join('\n'),
        },
      ];
    },
  };
}
