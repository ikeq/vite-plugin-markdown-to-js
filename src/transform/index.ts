export * from './imports';
export * from './vue';
