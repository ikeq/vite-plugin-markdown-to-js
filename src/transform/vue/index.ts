import kebabCase from 'lodash/kebabCase';
import { CodeBlock, LangTransform } from '../../parse';
import { parseSFC } from './sfc';


const defaultRenders: Record<'template' | 'script' | 'style', (code: string, meta: any, env: Parameters<LangTransform['transform']>[2]) => string> = {
  template: code => code.trim(),
  style: code => code.trim(),
  script: (code, { imports }) => {
    code = code.trim();
    if (!code) {
      return `export default { components: { ${imports || ''} } };`;
    }
    return code.replace('export default {', `export default {\n components: { ${imports || ''} },`);
  },
};

/**
 * @example
 * ```md
 * ```vue
 * <template>
 * </template>
 * <script>
 * export default {}
 * </script>
 * ```
 * ```
 */
export function transformVue(options: {
  importsAsComponents?: boolean | ((imports: string[]) => string);
  renders?: typeof defaultRenders
} = {}): LangTransform {
  return {
    lang: 'vue',
    tokenize(tokens) {
      if (!tokens.some((token) => token.type === 'code' && token.lang === 'vue')) {
        tokens.push({ type: 'code', lang: 'vue', text: '<script></script>', raw: '' });
      }

      return tokens;
    },
    transform(src, md, env) {
      const renders: typeof options.renders = {
        template: options.renders?.template || defaultRenders.template,
        style: options.renders?.style || defaultRenders.style,
        script: options.renders?.script || defaultRenders.script,
      };
      const blocks = parseSFC<'template' | 'script' | 'style'>(src);
      const ret: CodeBlock[] = [{
        type: 'html',
        code: md.replace('```vue', '```html'),
        transform: source => `<!-- langvue -->${source.replace(/^<pre>/, '<pre v-pre>')}<!-- endlangvue -->`,
      }];

      if (!blocks.length) return ret;

      const tags = blocks.reduce((map, i) => {
        return {
          ...map,
          [i.tag]: { ...i }
        };
      }, {} as Record<typeof blocks[0]['tag'], typeof blocks[0]>);

      if (!tags.script) {
        tags.script = {
          tag: 'script',
          code: '',
          attrs: '',
        };
      }
      if (!tags.template) {
        tags.template = {
          tag: 'template',
          code: '',
          attrs: '',
        };
      }
      if (!tags.style) {
        tags.style = {
          tag: 'style',
          code: '',
          attrs: '',
        };
      }

      Object.keys(tags).forEach((key) => {
        const tagName = key as keyof typeof tags;
        const block = tags[tagName];
        const renderer = renders[tagName];
        const code: CodeBlock['code'] = tagName === 'script' && options.importsAsComponents
          ? (scripts, html) => {
            const imports = scripts.map((i) => i.imports || []).flat(2);

            html.forEach((h) => {
              // consume imports block
              if (h.type === 'imports' && h.imports) {
                h.type = 'html';
                h.skipCompile = true;
                h.code =
                  typeof options.importsAsComponents === 'function'
                    ? options.importsAsComponents(h.imports)
                    : (h.imports.map((i: string) => `<${kebabCase(i)}/>`).join('') as string);
              }
            });

            return renderer(block.code, { imports }, env)
          }
          : renderer(block.code, {}, env);

        ret.unshift({ type: tagName, attrs: block.attrs, code, skipCompile: true });
      });

      return ret;
    },
  };
}
