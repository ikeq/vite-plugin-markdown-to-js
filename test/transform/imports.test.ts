import { transformImports } from '../../src/transform/imports';

describe('transformImports', () => {
  const env = { matter: {}, path: '' };

  it('basic', () => {
    const { transform } = transformImports();
    const raw = `a,a,b c \nd \nd`;
    expect(transform(raw, '```imports\na,b c \nd\n```', env)).toMatchSnapshot();
  });

  it('basic', () => {
    const { transform } = transformImports({ base: 'path/to/path', defaultPrefix: 'module' });
    const raw = `a,b c \nd`;
    expect(transform(raw, '```imports\na,b c \nd\n```', env)).toMatchSnapshot();
  });
});
