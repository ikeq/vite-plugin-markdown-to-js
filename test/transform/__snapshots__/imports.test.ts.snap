// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`transformImports basic 1`] = `
Array [
  Object {
    "code": "",
    "imports": Array [
      "a",
      "b",
      "c",
      "d",
    ],
    "type": "imports",
  },
  Object {
    "code": "import a from \\"a.md\\";
import b from \\"b.md\\";
import c from \\"c.md\\";
import d from \\"d.md\\";",
    "imports": Array [
      "a",
      "b",
      "c",
      "d",
    ],
    "type": "script",
  },
]
`;

exports[`transformImports basic 2`] = `
Array [
  Object {
    "code": "",
    "imports": Array [
      "moduleA",
      "moduleB",
      "moduleC",
      "moduleD",
    ],
    "type": "imports",
  },
  Object {
    "code": "import moduleA from \\"path/to/path/a.md\\";
import moduleB from \\"path/to/path/b.md\\";
import moduleC from \\"path/to/path/c.md\\";
import moduleD from \\"path/to/path/d.md\\";",
    "imports": Array [
      "moduleA",
      "moduleB",
      "moduleC",
      "moduleD",
    ],
    "type": "script",
  },
]
`;
